### PR TITLE
Add extra error handling to S3 remote logging (#9118)

### PR DESCRIPTION
**Pull request by JPonte**
https://github.com/apache/airflow/issues/9118
---
Make sure to mark the boxes below before creating PR: [x]

- [x] Description above provides context of the change
- [x] Unit tests coverage for changes (not needed for documentation changes)
- [x] Target Github ISSUE in description if exists
- [x] Commits follow "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)"
- [x] Relevant documentation is updated including usage instructions.
- [x] I will engage committers as explained in [Contribution Workflow Example](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#contribution-workflow-example).

---
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/master/UPDATING.md).
Read the [Pull Request Guidelines](https://github.com/apache/airflow/blob/master/CONTRIBUTING.rst#pull-request-guidelines) for more information.

### CODE REVIEW

**Comment by dimon222** [comment]
This works very well!

**Comment by RosterIn** [comment]
@feluelle can you check?

**Review by feluelle — APPROVED** [review]
LGTM 👍

**Comment by ashb** [comment]
Can you give me an example of what the log view now shows with an error?

**Comment by JPonte** [comment]
Depending on the type of the error it might show:
`*** Failed to verify remote log exists` if the HEAD to the s3 object fails along with the error.
If the HEAD returns 404 (meaning that the log file doesn't exist in s3) it shows:
`*** Falling back to local log`. 
And if the previous log to append fails to be read, it gets skipped.

**Comment by ashb** [comment]
Oh, so it doesn't include any of the error message from S3?

**Comment by JPonte** [comment]
It shows the error from the HEAD:
`'*** Failed to verify remote log exists {}.\n{}\n'.format(remote_loc, str(error))`
Shows the supposed location of the remote log and the error.

**Review by ashb — CHANGES_REQUESTED** [review]
Could you add a new test along the lines of `test_read_when_s3_log_missing` but when an error is returned.

**Review by ashb — APPROVED** [review]
Sorry, missed the rereview request.

Lgtm

**Comment by feluelle** [comment]
Sorry @JPonte but there are some conflicting files. Can you fix this? When you've done this I am merging it.

**Comment by JPonte** [comment]
@feluelle done

**Comment by feluelle** [comment]
@JPonte there is an error with the `Static checks`
```diff
-            msg = 'Could not read logs from {} with error: {}'.format(remote_log_location, error)
+            msg = f'Could not read logs from {remote_log_location} with error: {error}'
```

**Comment by JPonte** [comment]
@feluelle Ah, my bad. Fixed again.

File: tests/providers/amazon/aws/log/test_s3_task_handler.py [hunk 1]
```diff
@@ -143,12 +144,27 @@ def test_read_when_s3_log_missing(self):
...
+
     def test_read_raises_return_error(self):
         handler = self.s3_task_handler
         url = 's3://nonexistentbucket/foo'
         with mock.patch.object(handler.log, 'error') as mock_error:
             result = handler.s3_read(url, return_error=True)
-            msg = 'Could not read logs from %s' % url
+            msg = (
+                f'Could not read logs from {url} with error: An error occurred (NoSuchBucket) when '
+                f'calling the HeadObject operation: The specified bucket does not exist '
```
Comment (ashb) [inline tests/providers/amazon/aws/log/test_s3_task_handler.py]:
```suggestion
                f'calling the HeadObject operation: The specified bucket does not exist'
```

File: tests/providers/amazon/aws/log/test_s3_task_handler.py [hunk 2]
```diff
@@ -143,12 +144,27 @@ def test_read_when_s3_log_missing(self):
...
         self.assertEqual({'end_of_log': True}, metadata[0])
 
+    def test_s3_read_when_log_missing(self):
+        handler = self.s3_task_handler
+        url = 's3://bucket/foo'
+        with mock.patch.object(handler.log, 'error') as mock_error:
+            result = handler.s3_read(url, return_error=True)
+            msg = (
+                f'Could not read logs from {url} with error: An error occurred (404) when calling the '
+                f'HeadObject operation: Not Found '
```
Comment (ashb) [inline tests/providers/amazon/aws/log/test_s3_task_handler.py]:
```suggestion
                f'HeadObject operation: Not Found'
```

**Review by ashb — COMMENTED** [review]
Tests failing because of trailing spaces in strings

**Pull request merged**